app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,539 @@
+import android.view.MenuItem;
+import android.view.View;
+import android.widget.Toast;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.appcompat.app.AlertDialog;
+import androidx.appcompat.app.AppCompatActivity;
+import androidx.appcompat.widget.Toolbar;
+import androidx.fragment.app.DialogFragment;
+import androidx.fragment.app.Fragment;
+import androidx.fragment.app.FragmentTransaction;
+
+import org.kiwix.kiwixmobile.KiwixApplication;
+import org.kiwix.kiwixmobile.R;
+import org.kiwix.kiwixmobile.utils.SharedPreferenceUtil;
+
+import java.util.ArrayList;
+
+import javax.inject.Inject;
+
+/**
+ * Created by @Aditya-Sood as a part of GSoC 2019.
+ *
+ * This activity is the starting point for the module used for sharing zims between devices.
+ *
+ * The module is used for transferring ZIM files from one device to another, from within the
+ * app. Two devices are connected to each other using WiFi Direct, followed by file transfer.
+ *
+ * The module uses this activity along with {@link DeviceListFragment} to manage connection
+ * and file transfer between the devices.
+ * */
+public class LocalFileTransferActivity extends AppCompatActivity implements WifiP2pManager.ChannelListener, DeviceListFragment.DeviceActionListener {
+
+  public static final String TAG = "LocalFileTransferActvty"; // Not a typo, 'Log' tags have a length upper limit of 25 characters
+  public static final int REQUEST_ENABLE_WIFI_P2P = 1;
+  public static final int REQUEST_ENABLE_LOCATION_SERVICES = 2;
+  private static final int PERMISSION_REQUEST_CODE_COARSE_LOCATION = 1;
+  private static final int PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS = 2;
+
+  @Inject
+  SharedPreferenceUtil sharedPreferenceUtil;
+
+  private ArrayList<Uri> fileUriArrayList;  // For sender device, stores Uris of files to be transferred
+  private Boolean fileSendingDevice = false;// Whether the device is the file sender or not
+
+
+  /* Variables related to the WiFi P2P API */
+  private boolean wifiP2pEnabled = false; // Whether WiFi has been enabled or not
+  private boolean retryChannel = false;   // Whether channel has retried connecting previously
+
+  private WifiP2pManager manager;         // Overall manager of Wifi p2p connections for the module
+  private WifiP2pManager.Channel channel; // Connects the module to device's underlying Wifi p2p framework
+
+  private final IntentFilter intentFilter = new IntentFilter(); // For specifying broadcasts (of the P2P API) that the module needs to respond to
+  private BroadcastReceiver receiver = null; // For receiving the broadcasts given by above filter
+
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_local_file_transfer);
+    setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT); // Protect AsyncTask from orientation changes
+    KiwixApplication.getApplicationComponent().inject(this);
```

Comment:
So this is a bit more complicated now but for activity injection we have 2 methods to go about it `AndroidInjection` and `Regular`.
The main topic is how they supply `Activity` to injected classes. AndroidInjection can supply the specific activity very easily "LocalFileTransferActivity" in this case but if I wanted to just have "Activity" in my constructor I would need to write a specific mapping that downcasts LocalFileTransferActivity and you have to do this for all activities that inject this way.
The Regular way just forces you to be a lot more verbose with the injection as you supply the activity to it for use in constructor injection
```
KiwixApplication.getAplicationComponent().activityComponent()
            .activity(activity)
            .build()
            .inject(this)
```
vs
```
AndroidInjection.inject(this)
```
We need to align on our DI at some point but right now to inject a DIalogShower you will need to use regular injection or write a module for LocalFileTransferActivity to downcast it to Activity

Reply:
For each of the above two ways of injection and `KiwixApplication.getApplicationComponent().inject(this)`, I get the following error:
```
.../Kiwix/kiwix-android/app/src/main/java/org/kiwix/kiwixmobile/di/components/ApplicationComponent.java:46: error: [Dagger/MissingBinding] org.kiwix.kiwixmobile.utils.DialogShower cannot be provided without an @Provides-annotated method.
public interface ApplicationComponent {
       ^
  A binding with matching key exists in component: org.kiwix.kiwixmobile.di.components.ActivityComponent
      org.kiwix.kiwixmobile.utils.DialogShower is injected at
          org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity.dialogShower
      org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity is injected at
          org.kiwix.kiwixmobile.di.components.ApplicationComponent.inject(org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity)
```

In the regular-way suggested above, `.inject(this)` raised the following error: 
```
Cannot resolve method inject(org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity)
```

Any reason why my code isn't using the `ActivityModule.kt` file for the above regular injection?
And should I look into how to create that module to downcast LocalFileTransferActivity?

Reply:
Look more closely
What you wrote:
`KiwixApplication.getApplicationComponent().inject(this),`
What I wrote:
`KiwixApplication.getApplicationComponent().activityComponent().activity(this).build().inject(this)`

You need to create the ActivityComponent with the current activity so that an activity gets supplied to the constructor of AlertDialogShower

Reply:
Oh sorry, I need to pay more attention.
You need to create the method `inject(LocalFileTransferActivity)`, press ctrl+enter on the compile error in the class and select the create method option

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/544fcd2215f282e15c38b3c6aa2ace928da74949#diff-e1d59ae6e15dcb502e125fb6e243d15dR87)
I added the `injection()` in [ActivityComponent.kt](https://github.com/kiwix/kiwix-android/pull/1276/commits/544fcd2215f282e15c38b3c6aa2ace928da74949#diff-4588c0e3bd5e6f3762517a77ab6052abR40) and removed it from [ApplicationComponent.java](https://github.com/kiwix/kiwix-android/pull/1276/commits/544fcd2215f282e15c38b3c6aa2ace928da74949#diff-04e668036f8c6b4f4bf9c119d586e4b2L72)